### PR TITLE
fix: Normalize class names #300

### LINKED ISSUE #300: Prepended backslash in class name reloads the full metadata from scratch
### Bug Report

Follow up of https://github.com/symfony/symfony/issues/46837

#### Summary

When extracting class metadata, doctrine copies the original class data and caches it only the first time
But following that copy, calls to `addEntityListener` are made which are not repercuted on the alias https://github.com/doctrine/persistence/blob/25ec98a8cbd1f850e60fdb62c0ef77c162da8704/src/Persistence/Mapping/AbstractClassMetadataFactory.php?rgh-link-date=2022-07-04T14%3A09%3A36Z#L180

#### Solution

~~The alias should not be copied, but should just be added to an alias map to retrieve the original cached class straight away~~

### PR DESCRIPTION
**Pull request by Tofandel**
Fixes #300

### CODE REVIEW

File: src/Persistence/Mapping/AbstractClassMetadataFactory.php [hunk 1]
```diff
@@ -153,6 +153,14 @@ abstract protected function initializeReflection(
      */
     abstract protected function isEntity(ClassMetadata $class);
 
+    /**
+     * Removes the prepended backslash of a class string to conform with how php outputs class names
+     */
+    protected function normalizeClassName(string $className): string
```
Comment (greg0ire) [inline src/Persistence/Mapping/AbstractClassMetadataFactory.php]: This method should be `private` or at least `final` if you think it should be used from the children classes.

File: src/Persistence/Mapping/AbstractClassMetadataFactory.php [hunk 2]
```diff
@@ -153,6 +153,24 @@ abstract protected function initializeReflection(
      */
     abstract protected function isEntity(ClassMetadata $class);
 
+    /**
+     * Removes the prepended backslash of a class string to conform with how php outputs class names
+     *
+     * @psalm-param string $className
```
Comment (greg0ire) [inline src/Persistence/Mapping/AbstractClassMetadataFactory.php]: Isn't the input supposed to be a `class-string` as well?

File: src/Persistence/Mapping/AbstractClassMetadataFactory.php [hunk 3]
```diff
@@ -153,6 +153,24 @@ abstract protected function initializeReflection(
...
+     *
+     * @psalm-return class-string
+     *
+     * @psalm-suppress MoreSpecificReturnType
```
Comment (greg0ire) [inline src/Persistence/Mapping/AbstractClassMetadataFactory.php]: We centralize that kind of stuff here: https://github.com/doctrine/persistence/blob/3.0.x/psalm.xml

File: tests/Persistence/Mapping/AbstractClassMetadataFactoryTest.php [hunk 4]
```diff
@@ -78,6 +78,22 @@ public function testItThrowsWhenAttemptingToCheckTransientForShortAlias(): void
...
     }
+
+    public function testItGetsTheSameMetadataForBackslashedClassName(): void
+    {
+        $cmf = $this->getMockForAbstractClass(AbstractClassMetadataFactory::class);
+        $cmf
+            ->method('newClassMetadataInstance')
+            ->with(SomeOtherEntity::class)
+            ->willReturn(
+                $this->createMock(ClassMetadata::class)
```
Comment (greg0ire) [inline tests/Persistence/Mapping/AbstractClassMetadataFactoryTest.php]:
```suggestion
                $this->createStub(ClassMetadata::class)
```

**Comment by greg0ire** [comment]
If you are having static analysis issues, consider cloning this for Psalm: https://github.com/phpstan/phpstan/issues/7483

File: src/Persistence/Mapping/AbstractClassMetadataFactory.php (hunk 3 above)
Comment (Tofandel) [reply src/Persistence/Mapping/AbstractClassMetadataFactory.php]: Isn't that for global/file suppression only? I don't see how to suppress only for a method in that file, I wouldn't want to suppress this for the rest of the file as this is just a false negative for this particular method and this warning is usually useful

File: src/Persistence/Mapping/AbstractClassMetadataFactory.php (hunk 2 above)
Comment (Tofandel) [reply src/Persistence/Mapping/AbstractClassMetadataFactory.php]: Indeed

File: src/Persistence/Mapping/AbstractClassMetadataFactory.php (hunk 3 above)
Comment (Tofandel) [reply src/Persistence/Mapping/AbstractClassMetadataFactory.php]: I guess I could try adding a stub file as a workaround since I managed to fix this issue in psalm, the stub file would just need this

```
/**
 * @psalm-pure
 *
 * @return ($string is class-string ? ($characters is '\\' ? class-string : string) : string)
 *
 * @psalm-flow ($string) -> return
 */
function ltrim(string $string, string $characters = " \t\n\r\0\x0B") : string {}
```

Comment (Tofandel) [reply src/Persistence/Mapping/AbstractClassMetadataFactory.php]: Can be removed if https://github.com/vimeo/psalm/pull/8219 is accepted

File: psalm.phpstub [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<?php
+/**
+ * @psalm-pure
+ *
+ * @return ($string is class-string ? ($characters is '\\' ? class-string : string) : string)
+ *
+ * @psalm-flow ($string) -> return
+ */
+function ltrim(string $string, string $characters = " \t\n\r\0\x0B") : string {}
```
Comment (greg0ire) [inline psalm.phpstub]: Please configure your IDE to always add a newline at end of file, this will make the red pictogram disappear.

File: psalm.phpstub [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<?php
+/**
+ * @psalm-pure
```
Comment (greg0ire) [inline psalm.phpstub:6]: Can you please add a link to your PR in that docblock? :pray: 

Thanks a lot by the way, this is awesome, it will fix the issue for many Doctrine packages :)

File: psalm.phpstub (hunk 5 above)
Comment (Tofandel) [reply psalm.phpstub]: It would actually be better to add a `.editorconfig` with the project coding standards so that it's automatic, otherwise I'd have to change them for each project ;)
https://github.com/TYPO3/coding-standards/blob/main/.editorconfig

It would cover trailing whitespaces in blank lines, end of file line return, indent size etc etc

Quite an important file to have for a project with multiple people contributing

Comment (Tofandel) [reply psalm.phpstub]: Want me to add it to this PR?

File: psalm.phpstub (hunk 6 above)
Comment (Tofandel) [reply psalm.phpstub:6]: Done, no problem happy to help 😊

File: psalm.phpstub (hunk 5 above)
Comment (greg0ire) [reply psalm.phpstub]: I think it could be a good idea, but please make it a separate PR because I feel it's going to be controversial.

Comment (Tofandel) [reply psalm.phpstub]: Okay, then this PR should be good to go

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Note for maintainers or @Tofandel : I think the commits can be squashed.

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @Tofandel !

**Comment by writ3it** [comment]
Hi,

this PR breaks test ClassMetadataFactoryTest::testHasGetMetadataNamespaceSeparatorIsNotNormalized in doctrine/orm:2.12.x. 

You can check test results here: https://github.com/doctrine/orm/runs/7698594303?check_suite_focus=true.
Test was implemented here: https://github.com/doctrine/orm/commit/884131e51fa7f14fd30aaed0e0d226dcbf79c51b to solve issue described here: https://github.com/doctrine/orm/issues/1750. 

@Tofandel could you check this case? What we can do?

**Comment by Tofandel** [comment]
The threads are very hard to follow, but to me it seems, as per [the last comment](https://github.com/doctrine/orm/issues/1750#issuecomment-162355602), that the linked issue had nothing to do with class normalization, and was related to namespace and a different fix was applied, so it's safe to modify this test to isSame and isTrue instead of isNotSame and isFalse, because the test mentionned was a revert and does nothing to check the issue (unless I didn't understand it in the first place), and only checks for the opposite of what is expected, since `\DoctrineGlobal_Article` is in fact the same class as `DoctrineGlobal_Article`, which is an issue

You can even see that the real tests for the issues are marked as 
```
    /**
     * @group DDC-115
     */
```
But this specific test is not

And that this test was first created in https://github.com/doctrine/orm/commit/120e6942e83c3cfb4adc7798ddfc6bdd73554bc5
And then the next commit just modify this test to check for the exact opposite, which is not the desired behavior, but at the time it was for "performance concerns"

Given that compared to php 5, php 8 is ~120% faster and that doctrine has a lot more caching now and that this really is such a fast triming operation, it's not a valid argument anymore

**Comment by writ3it** [comment]
Thank you a lot for the explanation and improvement!
